pkg/reconciler/v1alpha1/trigger/trigger.go

```diff
@@ -0,0 +1,638 @@
+			return nil, err
+		}
+		return sub, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	// Update Subscription if it has changed. Ignore the generation.
+	expected.Spec.DeprecatedGeneration = sub.Spec.DeprecatedGeneration
+	if !equality.Semantic.DeepDerivative(expected.Spec, sub.Spec) {
+		// Given that the backing channel spec is immutable, we cannot just update the subscription.
+		// We delete it instead, and re-create it.
+		err = r.client.Delete(ctx, sub)
+		if err != nil {
+			logging.FromContext(ctx).Info("Cannot delete subscription", zap.Error(err))
+			r.recorder.Eventf(t, corev1.EventTypeWarning, subscriptionDeleteFailed, "Delete Trigger's subscription failed: %v", err)
+			return nil, err
+		}
+		sub = expected
+		err = r.client.Create(ctx, sub)
+		if err != nil {
+			logging.FromContext(ctx).Info("Cannot create subscription", zap.Error(err))
+			r.recorder.Eventf(t, corev1.EventTypeWarning, subscriptionCreateFailed, "Create Trigger's subscription failed: %v", err)
+			return nil, err
+		}
+	}
+	return sub, nil
+}
+
+// getSubscription returns the subscription of trigger 't' if exists,
+// otherwise it returns an error.
+func (r *reconciler) getSubscription(ctx context.Context, t *v1alpha1.Trigger) (*v1alpha1.Subscription, error) {
+	list := &v1alpha1.SubscriptionList{}
+	opts := &runtimeclient.ListOptions{
+		Namespace:     t.Namespace,
+		LabelSelector: labels.SelectorFromSet(subscriptionLabels(t)),
+		// Set Raw because if we need to get more than one page, then we will put the continue token
+		// into opts.Raw.Continue.
+		Raw: &metav1.ListOptions{},
+	}
+
+	err := r.client.List(ctx, opts, list)
+	if err != nil {
+		return nil, err
+	}
+	for _, s := range list.Items {
+		if metav1.IsControlledBy(&s, t) {
+			return &s, nil
+		}
+	}
+
+	return nil, k8serrors.NewNotFound(schema.GroupResource{}, "")
+}
+
+// makeSubscription returns a placeholder subscription for trigger 't', channel 'c', and service 'svc'.
+func makeSubscription(t *v1alpha1.Trigger, c *v1alpha1.Channel, svc *corev1.Service) *v1alpha1.Subscription {
+	return &v1alpha1.Subscription{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    t.Namespace,
+			GenerateName: fmt.Sprintf("%s-%s-", t.Spec.Broker, t.Name),
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(t, schema.GroupVersionKind{
+					Group:   v1alpha1.SchemeGroupVersion.Group,
+					Version: v1alpha1.SchemeGroupVersion.Version,
+					Kind:    "Trigger",
+				}),
+			},
+			Labels: subscriptionLabels(t),
+		},
+		Spec: v1alpha1.SubscriptionSpec{
+			Channel: corev1.ObjectReference{
+				APIVersion: v1alpha1.SchemeGroupVersion.String(),
+				Kind:       "Channel",
+				Name:       c.Name,
+			},
+			Subscriber: &v1alpha1.SubscriberSpec{
+				Ref: &corev1.ObjectReference{
+					APIVersion: "v1",
+					Kind:       "Service",
+					Name:       svc.Name,
+				},
+			},
+			// TODO This pushes directly into the Channel, it should probably point at the Broker ingress instead.
```

Comment:
Pushing reply events directly into the Broker's channel bypasses any filtering or other functionality of the Broker ingress. One solution is creating a separate Channel to receive replies, plus a Subscription from that Channel to the Broker's ingress.

Reply:
Tracked in #905

Reply:
Also, the existing code doesn't push back any replies, because the message dispatcher library we are using doesn't return anything, so nothing was ever replied.

Reply:
> the existing code doesn't push back any replies

Tracked in #906